File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let maxLength = 0;
    let stack = [-1]; // Base for calculating the length

    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i);
        } else {
            stack.pop();
            if (stack.length === 0) {
                stack.push(i); // Update the base for length calculation
            } else {
                maxLength = Math.max(maxLength, i - stack[stack.length - 1]);
            }
        }
    }
    return maxLength;
};

console.log(longestValidParentheses("(()")); // 2
console.log(longestValidParentheses(")()())")); // 4
console.log(longestValidParentheses("")); // 0

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.go
func longestValidParentheses(s string) int {
    maxLength := 0
    stack := []int{-1} // Base for calculating the length

    for i := 0; i < len(s); i++ {
        if s[i] == '(' {
            stack = append(stack, i)
        } else {
            stack = stack[:len(stack)-1]
            if len(stack) == 0 {
                stack = append(stack, i) // Update the base for length calculation
            } else {
                maxLength = max(maxLength, i-stack[len(stack)-1])
            }
        }
    }
    return maxLength
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
